src/coreclr/System.Private.CoreLib/src/System/Variant.cs

```diff
@@ -78,323 +23,311 @@ internal partial struct Variant
-        internal extern void SetFieldsObject(object val);
-
-        //
-        // Constructors
-        //
-
-        internal Variant(int flags, object or, long data)
-        {
-            _flags = flags;
-            _objref = or;
-            _data = data;
-        }
-
-        public Variant(bool val)
-        {
-            _objref = null;
-            _flags = CV_BOOLEAN;
-            _data = (val) ? bool.True : bool.False;
-        }
-
-        public Variant(sbyte val)
-        {
-            _objref = null;
-            _flags = CV_I1;
-            _data = val;
-        }
-
-        public Variant(byte val)
-        {
-            _objref = null;
-            _flags = CV_U1;
-            _data = val;
-        }
-
-        public Variant(short val)
-        {
-            _objref = null;
-            _flags = CV_I2;
-            _data = val;
-        }
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Variant_ConvertValueTypeToRecord")]
+        private static partial void ConvertValueTypeToRecord(ObjectHandleOnStack obj, out ComVariant pOle);
 
-        public Variant(ushort val)
+        internal static ComVariant GetComIPFromObjectRef(object? obj)
         {
-            _objref = null;
-            _flags = CV_U2;
-            _data = val;
+            IntPtr pUnk = GetIUnknownOrIDispatchForObject(ObjectHandleOnStack.Create(ref obj), out bool isIDispatch);
+            return ComVariant.CreateRaw(isIDispatch ? VarEnum.VT_DISPATCH : VarEnum.VT_UNKNOWN, pUnk);
         }
 
-        public Variant(char val)
-        {
-            _objref = null;
-            _flags = CV_CHAR;
-            _data = val;
-        }
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "MarshalNative_GetIUnknownOrIDispatchForObject")]
+        private static partial IntPtr GetIUnknownOrIDispatchForObject(ObjectHandleOnStack o, [MarshalAs(UnmanagedType.Bool)] out bool isIDispatch);
 
-        public Variant(int val)
+        private static object? GetObjectRefFromComIP(IntPtr pUnk)
         {
-            _objref = null;
-            _flags = CV_I4;
-            _data = val;
+            return pUnk == IntPtr.Zero ? null : Marshal.GetObjectForIUnknown(pUnk);
         }
 
-        public Variant(uint val)
+        private static unsafe object? ConvertWrappedObject(object? wrapped) => wrapped switch
         {
-            _objref = null;
-            _flags = CV_U4;
-            _data = val;
-        }
-
-        public Variant(long val)
+            null or Empty => null,
+            Enum => wrapped.GetType(),
+            IntPtr or UIntPtr => wrapped,
+            ValueType v when RuntimeHelpers.GetMethodTable(v)->IsPrimitive => null,
+            DateTime or TimeSpan or Currency => null,
+            _ => wrapped,
+        };
+
+        // Helper code for marshaling managed objects to VARIANT's
+        internal static void MarshalHelperConvertObjectToVariant(object? o, out ComVariant pOle)
```

Comment:
Any chance we can update this to instead of an `out` we return the `ComVariant`?

Reply:
`MethodDescCallSite::CallWithValueTypes_RetArgSlot`? It may be more complex to copy the result at native side.

Reply:
I'll look at this later.